mlflow/tracing/trace_manager.py

```diff
@@ -0,0 +1,177 @@
+import logging
+import threading
+from dataclasses import dataclass, field
+from functools import lru_cache
+from typing import Dict, Optional
+
+from opentelemetry import trace as trace_api
+
+from mlflow.tracing.types.model import Status, StatusCode, Trace, TraceData, TraceInfo
+from mlflow.tracing.types.wrapper import MLflowSpanWrapper, NoOpMLflowSpanWrapper
+
+_logger = logging.getLogger(__name__)
+
+
+# Internal representation to keep the state of a trace.
+# Dict[str, Span] is used instead of TraceData to allow access by span_id.
+@dataclass
+class _Trace:
+    trace_info: TraceInfo
+    span_dict: Dict[str, MLflowSpanWrapper] = field(default_factory=dict)
+
+    def to_mlflow_trace(self) -> Trace:
+        span_list = [span.to_mlflow_span() for span in self.span_dict.values()]
+        return Trace(self.trace_info, TraceData(spans=span_list))
+
+
+class InMemoryTraceManager:
+    """
+    Manage spans and traces created by the tracing system in memory.
+    """
+
+    _instance_lock = threading.Lock()
+    _instance = None
+
+    @classmethod
+    def get_instance(cls):
+        if cls._instance is None:
+            with cls._instance_lock:
+                if cls._instance is None:
+                    cls._instance = InMemoryTraceManager()
+        return cls._instance
+
+    def __init__(self):
+        self._traces: Dict[str, _Trace] = {}
+        self._lock = threading.Lock()  # Lock for _traces
+
+    def start_detached_span(
+        self,
+        name: str,
+        trace_id: Optional[str] = None,
+        parent_span_id: Optional[str] = None,
+        span_type: Optional[str] = None,
+    ) -> MLflowSpanWrapper:
+        """
+        Start a new OpenTelemetry span that is not part of the current trace context, but with the
+        explicit parent span ID if provided.
+
+        Args:
+            name: The name of the span.
+            trace_id: The trace ID for the span. Only used for getting the parent span for the given
+                parent_span_id. If not provided, a new trace will be created.
+            parent_span_id: The parent span ID of the span. If None, the span will be a root span.
+            span_type: The type of the span.
+
+        Returns:
+            The newly created span (wrapped in MLflowSpanWrapper). If any error occurs, returns a
+            NoOpMLflowSpanWrapper that has exact same interface but no-op implementations.
+        """
+        from mlflow.tracing.provider import get_tracer
+
+        try:
+            tracer = get_tracer(__name__)
+            if parent_span_id:
+                parent_span = self.get_span_from_id(trace_id, parent_span_id)._span
+                context = trace_api.set_span_in_context(parent_span)
+            else:
+                context = None
+
+            span = MLflowSpanWrapper(tracer.start_span(name, context=context), span_type=span_type)
```

Comment:
How do we make sure the start_time and end_time are correctly converted to int (and in microseconds) in MLflow span? (not a comment for this line though)

Reply:
Ah true, the timestamp is not converted to microsecond now. I think doing division in start/end_time property and to_mlflow_span method would be sufficient.